src/Data/Equality/Graph/Classes.hs

```diff
@@ -27,9 +29,9 @@ data EClass l = EClass
     { eClassId      :: {-# UNPACK #-} !ClassId -- ^ E-class identifier
     , eClassNodes   :: !(S.Set (ENode l))      -- ^ E-nodes in this class
     , eClassData    :: Domain l                -- ^ The analysis data associated with this eclass.
-    , eClassParents :: !(NodeMap l ClassId)    -- ^ E-nodes which are parents of this e-class and their corresponding e-class ids. We found a mapping from nodes to e-class ids a better representation than @[(ENode l, ClassId)]@, and we get de-duplication built-in.
+    , eClassParents :: !(SList (ClassId, ENode l))   -- ^ E-nodes which are parents of this e-class and their corresponding e-class ids.
```

Comment:
I'm curious what the length is used for. Is it really worth an extra indirection?

Are we going to append this list to the worklist when we merge this class? Then maybe we would like a data structure that appends better? On the other hand, we do have to traverse the whole structure anyway for deduping, so probably it doesn't matter. This requires a finer understanding of the data structure than I have :slightly_smiling_face: .